llvm/include/llvm/Analysis/GenericDomTreeUpdater.h

```diff
@@ -146,8 +152,25 @@ class GenericDomTreeUpdater {
   /// 2. It is illegal to submit any update that has already been submitted,
   /// i.e., you are supposed not to insert an existent edge or delete a
   /// nonexistent edge.
+  /// 3. This kind updates are incompatible with critical edge splitting
+  /// updates, call this method will apply all critical edge updates in
+  /// lazy mode. It is not recommended to interleave applyUpdates and
+  /// applyUpdatesForCriticalEdgeSplitting.
```

Comment:
I can't parse this sentence. Could you explain it in a bit more detail?

Specifically, it is not clear to me how to reconcile 'incompatible' with 'not recommended'. What happens if we do add edge splitting updates in eager and lazy mode?

Reply:
What do you think about exposing a more general update type that captures both edge additions/deletions and critical edge splitting? Would that make sense? I'm trying to see if we can avoid adding new methods and, in turn, having to document intricacies like this.

Reply:
From user perspective a unified update method is better, i.e. `applyUpdates({{From1, To1, UpdateKind::Insert}, {From, To, New, UpdateKind::Split}...})`, then `cfg::Update<NodePtrT>` is not enough.

Reply:
> then cfg::Update<NodePtrT> is not enough.

Could you explain this part? I'm not sure which API we are talking about and what the limitations are.

Reply:
Currently all update operations are based on [`cfg::Update`](https://llvm.org/doxygen/classllvm_1_1cfg_1_1Update.html). If we extend `cfg::Update` directly e.g. add a new kind `UpdateKind::Split`, then it is duplicated with two insertions and deletion, and causes some compiler warnings.

Reply:
We could update `cfg::Update` though and add one more node ref there. I still don't understand what the issue is here:
> then it is duplicated with two insertions and deletion, and causes some compiler warnings.

Reply:
A critical splitting update can be expressed as insert two new edges and delete old critical edge.
Current CFG codes rely on the dichotomy of `UpdateKind`, a new update kind makes old match not exhaustive.
Some CFG related codes need to be updated, e.g. `LegalizeUpdates` which is not easy to update, because it assumes `cfg::Update` only has two pointer fields...